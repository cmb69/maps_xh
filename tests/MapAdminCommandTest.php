<?php

namespace Maps;

use ApprovalTests\Approvals;
use Maps\Model\Map;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\CsrfProtector;
use Plib\DocumentStore2 as DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class MapAdminCommandTest extends TestCase
{
    private DocumentStore $store;
    /** @var CsrfProtector&Stub */
    private $csrfProtector;
    private View $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $map = Map::create("london", $this->store);
        $map->addMarker(0, 0, "basic info", true);
        $this->store->commit();
        $this->csrfProtector = $this->createStub(CsrfProtector::class);
        $this->csrfProtector->method("token")->willReturn("0123456789ABCDEF");
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["maps"]);
    }

    private function sut(): MapAdminCommand
    {
        return new MapAdminCommand("./", $this->store, $this->csrfProtector, $this->view);
    }

    public function testRendersOverview(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()($request);
        $this->assertSame("Maps – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersEditorForNewMap(): void
    {
        $request = new FakeRequest(["url" => "http://example.com/?&maps&admin=plugin_main&action=create"]);
        $response = $this->sut()($request);
        $this->assertSame("Maps – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testCreatesNewMap(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=create",
            "post" => [
                "maps_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertFileExists(vfsStream::url("root/new.xml"));
        $this->assertSame("http://example.com/?&maps&admin=plugin_main", $response->location());
    }

    public function testCreatingIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=create",
            "post" => [
                "maps_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToSaveNewMap(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=create",
            "post" => [
                "maps_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the map!", $response->output());
    }

    public function testRendersEditorForUpdate(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
        ]);
        $response = $this->sut()($request);
        $this->assertSame("Maps – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testsReportsThatNoMapIsSelected(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You have not selected a map!", $response->output());
    }

    public function testReportsMissingMap(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=does-not-exist",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the map “does-not-exist”!", $response->output());
    }

    public function testUpdatesMap(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
            "post" => [
                "maps_do" => "",
                "markers" => '[{"latitude":0,"longitude":0,"info":"more info","show":false}]',
            ],
        ]);
        $response = $this->sut()($request);
        $map = Map::read("london", $this->store);
        $this->assertCount(1, $map->markers());
        $this->assertSame("http://example.com/?&maps&admin=plugin_main&maps_map=london", $response->location());
    }

    public function testsReportsMissingMapWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=does-not-exist",
            "post" => [
                "maps_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the map “does-not-exist”!", $response->output());
    }

    public function testsReportsThatNoMapIsSelectedWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update",
            "post" => [
                "maps_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You have not selected a map!", $response->output());
    }

    public function testUpdateIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
            "post" => [
                "maps_do" => "",
                "markers" => "0|0|more info|",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToUpdateMap(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=update&maps_map=london",
            "post" => [
                "maps_do" => "",
                "markers" => "0|0|more info|",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the map!", $response->output());
    }

    public function testRendersImportForm(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=london",
        ]);
        $response = $this->sut()($request);
        $this->assertSame("Maps – Import", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testsReportsThatNoMapIsSelectedForImport(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You have not selected a map!", $response->output());
    }

    public function testReportsMissingMapForImport(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=does-not-exist",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the map “does-not-exist”!", $response->output());
    }

    public function testImportsGeoJson(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=london",
            "post" => [
                "maps_do" => "",
                "geojson" => '{"features":[{"properties":{"foo":"bar"},'
                    . '"geometry":{"type":"Point","coordinates":[0,0]}}]}',
                "template" => "{foo}",
            ],
        ]);
        $response = $this->sut()($request);
        $map = Map::read("london", $this->store);
        $this->assertCount(2, $map->markers());
        $this->assertSame("http://example.com/?&maps&admin=plugin_main&maps_map=london", $response->location());
    }

    public function testsReportsThatNoMapIsSelectedWhenImporting(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import",
            "post" => [
                "maps_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You have not selected a map!", $response->output());
    }

    public function testsReportsMissingMapWhenImporting(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=does-not-exist",
            "post" => [
                "maps_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the map “does-not-exist”!", $response->output());
    }

    public function testImportIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=london",
            "post" => [
                "maps_do" => "",
                "geojson" => '{}',
                "template" => "{foo}",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testsReportsUnsupportedGeoJson(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=london",
            "post" => [
                "maps_do" => "",
                "geojson" => '{}',
                "template" => "{foo}",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("This GeoJSON is not supported!", $response->output());
    }

    public function testReportsFailureToImport(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&maps&admin=plugin_main&action=import&maps_map=london",
            "post" => [
                "maps_do" => "",
                "geojson" => '{"features":[{"properties":{"foo":"bar"},'
                    . '"geometry":{"type":"Point","coordinates":[0,0]}}]}',
                "template" => "{foo}",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the map!", $response->output());
    }
}
