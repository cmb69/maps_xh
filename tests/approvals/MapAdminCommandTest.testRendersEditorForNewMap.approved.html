
<script type="module" src="./admin.js"></script>
<article class="maps_edit">
  <h1>Maps â€“ Edit</h1>
  <form method="post">
    <label>
      <span>Name</span>
      <span class="maps_help">(only a-z, 0-9, and hyhpens)</span>
      <input name="name" value=""  required pattern="[a-z0-9\-]+">
    </label>
    <label>
      <span>Title</span>
      <input name="title" value="" required>
    </label>
    <label>
      <span>Latitude</span>
      <span class="maps_help">(between -90 and 90)</span>
      <input type="number" name="latitude" value="0" min="-90" max="90" step="any">
    </label>
    <label>
      <span>Longitude</span>
      <span class="maps_help">(between -180 and 180)</span>
      <input type="number" name="longitude" value="0" min="-180" max="180" step="any">
    </label>
    <label>
      <span>Zoom</span>
      <span class="maps_help">(between 0 and 20)</span>
      <input type="number" name="zoom" value="0" min="0" max="20">
    </label>
    <label>
      <span>Max. Zoom</span>
      <span class="maps_help">(between 0 and 20)</span>
      <input type="number" name="max_zoom" value="0" min="0" max="20">
    </label>
    <label>
      <span>Aspect Ratio</span>
      <span class="maps_help">(with/height)</span>
      <input name="aspect_ratio" value="1/1" required pattern="\d+/\d+">
    </label>
    <label>
      <span>Markers</span>
      <textarea name="markers"></textarea>
    </label>
    <table style="display: none">
      <caption>Markers</caption>
      <colgroup>
        <col>
        <col>
        <col>
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th>Latitude <span class="maps_help">(between -90 and 90)</span></th>
          <th>Longitude <span class="maps_help">(between -180 and 180)</span></th>
          <th>Info <span class="maps_help">(arbitrary HTML)</span></th>
          <th></th>
          <th><button class="maps_add_row" type="button"><svg xmlns="http://www.w3.org/2000/svg" role="img" viewBox="0 0 384 512" width="1em" height="1em" fill="currentColor"><title>Create marker</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"/></svg></button></th>
        </tr>
      </thead>
      <tbody>
      </tbody>
      <tfoot style="display: none">
        <tr>
          <td><input type="number" value="0" min="-90" max="90" step="any"></td>
          <td><input type="number" value="0" min="-180" max="180" step="any"></td>
          <td><textarea></textarea></td>
          <td><input type="checkbox"></td>
          <td><button type="button" class="maps_delete_row"><svg xmlns="http://www.w3.org/2000/svg" role="img" viewBox="0 0 448 512" width="1em" height="1em" fill="currentColor"><title>Delete this marker</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"/></svg></button></td>
        </tr>
      </tfoot>
    </table>
    <p class="maps_controls">
      <button name="maps_do">Save</button>
    </p>
    <input type="hidden" name="maps_token" value="0123456789ABCDEF">
  </form>
</article>
